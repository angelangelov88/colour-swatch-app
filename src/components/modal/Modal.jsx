import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Dialog, Transition } from '@headlessui/react'
import { Fragment, useState } from 'react'
import { XIcon } from '@heroicons/react/outline'
import { Button } from '../button/Button'
import { InfoText } from './InfoText'

const Modal = (props) => {
  let {
    children,
    className,
    title,
    infoTextColor,
    infoTextChildren,
  } = props

  const [open, setOpen] = useState(false)

  function closeModal() {
    setOpen(false)
  }

  function openModal() {
    setOpen(true)
  }

  const classes = classNames(
    "absolute top-[calc(100vh-90%)] w-[703px] h-[527px] font-ibm flex flex-col py-6 bg-white border border-gray-20 rounded-lg shadow-2xl",
    open ? "block" : "hidden",
    className
  )

  return (
    <>
      <div onClick={openModal}>Open Modal</div>
      <Transition appear show={open} as={Fragment}>
        <Dialog className="flex justify-center" open={open} onClose={closeModal}>
          <Transition.Child
            as={Fragment}
            enter="ease-out duration-300"
            enterFrom="opacity-0"
            enterTo="opacity-100"
            leave="ease-in duration-200"
            leaveFrom="opacity-100"
            leaveTo="opacity-0"
          >
            <div className="fixed inset-0 bg-black bg-opacity-25" />
          </Transition.Child>
          <Dialog.Panel className={classes}>
            <Dialog.Title>
              <p className="text-3xl font-medium px-6">{title}</p>
              <span onClick={closeModal}>
                <XIcon 
                  className="absolute top-6 right-6 w-7 h-7 text-gray-50 hover:cursor-pointer" 
                />
              </span>
            </Dialog.Title>
            <InfoText 
              color={infoTextColor} 
              children={infoTextChildren}
            />
            <div className="h-[20rem] mb-6">
              {children}
            </div>
            <div className="flex justify-end pr-6">
              <Button 
                children="Cancel"
                className="mr-6"
                variant="secondary"
                rounded="lg"
                tabIndex="0"
                onClick={closeModal}
              />
              <Button 
                children="Continue"
                rounded="lg"
                tabIndex="0"
              />
            </div>
          </Dialog.Panel>
        </Dialog>
      </Transition>
    </>
  )
}

Modal.defaultProps = {
  children: undefined,
  title: "Title",
  infoTextColor: "primary",
  infoTextChildren: "There are a few ways to get started with a new document in Legalesign. Start by telling us what youâ€™d like to do:",
}

Modal.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  title: PropTypes.string,
  infoTextColor: PropTypes.oneOf([
    "primary",
    "green",
    "yellow",
    "red",
  ]),
  infoTextChildren: PropTypes.node,
}

export { Modal }
