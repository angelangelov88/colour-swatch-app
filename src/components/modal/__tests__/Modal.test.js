import React from 'react';
import renderer from 'react-test-renderer';
import Modal from '../Modal';

it('should render with all default settings', () => {
  const component = renderer.create(
    <Modal />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with children just as text and the rest default', () => {
  const component = renderer.create(
    <Modal>Hello there</Modal>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with red colour', () => {
  const component = renderer.create(
    <Modal
      color="red"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with yellow colour', () => {
  const component = renderer.create(
    <Modal
      color="yellow"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with primary colour', () => {
  const component = renderer.create(
    <Modal
      color="primary"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with green colour', () => {
  const component = renderer.create(
    <Modal
      color="green"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with info text and heading in it saying: Test which is bold and warning text here', () => {
  const component = renderer.create(
    <Modal
      text={(
        <p>
          <span className="font-medium">Test:</span>
          {' '}
          Warning text here
        </p>
    )}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with text Warning! which is font-weight 500 and no more text', () => {
  const component = renderer.create(
    <Modal
      text={
        <p><span className="font-medium">Warning:</span></p>
    }
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with text as a string', () => {
  const component = renderer.create(
    <Modal
      text="This is just some text"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with title Open your app', () => {
  const component = renderer.create(
    <Modal
      title="Open your app"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with open prop set to showDialog and onClose to handleClose', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((initialState) => [initialState, setState]);
  const [showDialog, setShowDialog] = React.useState(false);
  const handleClose = () => setShowDialog(false);
  const component = renderer.create(
    <Modal
      open={showDialog}
      onClose={handleClose}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with SubmitButton prop set as a plain button and onClick function handleConfirm', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((initialState) => [initialState, setState]);
  const [showDialog, setShowDialog] = React.useState(false);
  const handleConfirm = () => { alert(123); setShowDialog(false); };
  const component = renderer.create(
    <Modal
      SubmitButton={() => <button onClick={handleConfirm}>button</button>}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
