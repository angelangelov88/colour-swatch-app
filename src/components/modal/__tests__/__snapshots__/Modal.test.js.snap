// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render in green" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render in primary" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render in red" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render in yellow" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with BtnIcon bell" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with SubmitButton prop set as a plain button and onClick function handleConfirm 1`] = `null`;

exports[`should render with all default settings 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with btnText Again" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with children as html 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with children as text 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with children just as text and the rest default 1`] = `null`;

exports[`should render with function for the button which displays handled! in console" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with green colour 1`] = `null`;

exports[`should render with info text and heading in it saying: Test which is bold and warning text here 1`] = `null`;

exports[`should render with open prop set to showDialog and onClose to handleClose 1`] = `null`;

exports[`should render with primary colour 1`] = `null`;

exports[`should render with red colour 1`] = `null`;

exports[`should render with text Warning! which is font-weight 500 and no more text 1`] = `null`;

exports[`should render with text as a string 1`] = `null`;

exports[`should render with text in children" 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with text in children" 2`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with title 1`] = `
<div
  onClick={[Function]}
>
  Open Modal
</div>
`;

exports[`should render with title Open your app 1`] = `null`;

exports[`should render with yellow colour 1`] = `null`;
